### PR TITLE
Adding request id plugin to uniquely track requests in APISIX

### LINKED ISSUE #2022: feature: Request ID plugin to uniquely identify requests in APISIX
### Issue description
When a client requests an API error I want to find the relevant request in the access logs. Is there a way to provide unique id. for each request and append in the request header?

### PR DESCRIPTION
**Pull request by sshniro**
### What this PR does / why we need it:
Resolves #2022

### CODE REVIEW

File: apisix/plugins/request-id.lua [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.rewrite(conf)
+    local headers = ngx.req.get_headers()
+    uuid.seed()
```
Comment (membphis) [inline apisix/plugins/request-id.lua:46]: we have set the right `seed`, so we do not need to set it again:

https://github.com/apache/apisix/blob/master/apisix/init.lua#L71

let we remove it.

File: apisix/plugins/request-id.lua [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.rewrite(conf)
```
Comment (membphis) [inline apisix/plugins/request-id.lua:44]: `_M.rewrite(conf)` -> `_M.rewrite(conf, ctx)`

File: apisix/plugins/request-id.lua [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    local headers = ngx.req.get_headers()
+    uuid.seed()
+    local uuid_val = uuid()
+    if not headers[conf.header_name] then
+        core.request.set_header(conf.header_name, uuid_val)
+    end
+
+    if conf.include_in_response then
+        ngx.ctx.api_ctx.x_request_id = uuid_val
```
Comment (membphis) [inline apisix/plugins/request-id.lua:53]: `ctx.x_request_id = uuid_val`, simpler way

File: apisix/plugins/request-id.lua [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    end
+end
+
+
+function _M.header_filter(conf, ctx)
+    if conf.include_in_response then
```
Comment (membphis) [inline apisix/plugins/request-id.lua:59]: another style:

```lua
if not conf.include_in_response then
    return
end

local headers = ngx.resp.get_headers()
if not headers[conf.header_name] then
      core.response.set_header(conf.header_name, ctx.x_request_id)
end
```

File: apisix/plugins/request-id.lua [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+local core          = require("apisix.core")
+local plugin_name   = "request-id"
+local ngx           = ngx
+local uuid          = require("resty.jit-uuid")
+
+local schema = {
+    type = "object",
+    properties = {
+        header_name = {type = "string", default = "X-Request-Id"},
+        include_in_response = {type = "boolean", default = false}
```
Comment (membphis) [inline apisix/plugins/request-id.lua:26]: I think the default value should be `true`.

**Comment by moonming** [comment]
Is this PR for tracing like skywalking or zipkin?

YuanSheng Wang <notifications@github.com> 于 2020年8月9日周日 上午9:23写道：

**Comment by sshniro** [comment]
@moonming  the requirement is to provide customers and API consumers with a request ID to report any API failures. This will helps to easily track down the root cause. 

Many provide such id with each request, the following is a format provided by [AWS](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-id-values/)

**Comment by moonming** [comment]
got it, thanks for your explanation

Thanks,
Ming Wen
Twitter: _WenMing


Nirojan Selvanathan <notifications@github.com> 于2020年8月10日周一 下午3:36写道：

**Comment by sshniro** [comment]
@membphis small question why does consequent calls to the endpoint returns the same value to the header? However if I call this endpoint via curl I do get different results. Is this something related to UUID seed not getting set in unit tests?
```shell
=== TEST X: check for unique id
--- config
    location /t {
        content_by_lua_block {
            local http = require "resty.http"
            local httpc = http.new()
            local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/opentracing"
            local res1, err1 = httpc:request_uri(uri,
                {
                    method = "GET",
                    headers = {
                        ["Content-Type"] = "application/json",
                    }
                })
            local res2, err2 = httpc:request_uri(uri,
                {
                    method = "GET",
                    headers = {
                        ["Content-Type"] = "application/json",
                    }
                })

            if res1.headers["X-Request-Id"] ~= res2.headers["X-Request-Id"] then
                ngx.say("unique request id")
            else
                ngx.say("failed")
            end
        }
    }
--- request
GET /t
--- response_body
unique request id
--- no_error_log
[error]
```

File: apisix/plugins/request-id.lua (hunk 5 above)
Comment (sshniro) [reply apisix/plugins/request-id.lua:26]: agreed.

**Comment by membphis** [comment]
It works fine at my side. And we got different `request id` in test case `2`.

I think you can append those test cases.

```
=== TEST 1: enable request-id with default value
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/routes/1',
                 ngx.HTTP_PUT,
                 [[{
                        "plugins": {
                            "request-id": {
                            }
                        },
                        "upstream": {
                            "nodes": {
                                "127.0.0.1:1982": 1
                            },
                            "type": "roundrobin"
                        },
                        "uri": "/hello"
                }]]
                )

            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]



=== TEST 2: check for unique id
--- config
    location /t {
        content_by_lua_block {
            local http = require "resty.http"
            local httpc = http.new()
            local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/hello"
            local res1, err1 = httpc:request_uri(uri,
                {
                    method = "GET",
                    headers = {
                        ["Content-Type"] = "application/json",
                    }
                }
            )
            local res2, err2 = httpc:request_uri(uri,
                {
                    method = "GET",
                    headers = {
                        ["Content-Type"] = "application/json",
                    }
                }
            )

            -- ngx.say("res1: ", res1.headers["X-Request-Id"])
            -- ngx.say("res2: ", res2.headers["X-Request-Id"])
            if res1.headers["X-Request-Id"] == res2.headers["X-Request-Id"] then
                ngx.say("unique request id")
            else
                ngx.say("failed")
            end
        }
    }
--- request
GET /t
--- response_body
failed
--- no_error_log
[error]
```

**Comment by sshniro** [comment]
@membphis thanks, pending issues have been fixed now.

**Comment by membphis** [comment]
@sshniro nice PR, merged already

**Pull request merged**